### PR TITLE
WIP: Add machineconfiguration API

### PR DESCRIPTION
**Pull request by LorbusChris**

### CODE REVIEW

File: machineconfiguration/OWNERS [hunk 1]
```diff
@@ -0,0 +1,3 @@
+reviewers:
+  - runcom
+  - kikisdeliveryservice
```
Comment (runcom) [inline machineconfiguration/OWNERS:3]: add yourself and Colin as well ;)

File: machineconfiguration/v1/register.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+// Resource generated code relies on this being here, but it logically belongs to the group
+// DEPRECATED
+func Resource(resource string) schema.GroupResource {
+	return schema.GroupResource{Group: GroupName, Resource: resource}
+}
+
+// addKnownTypes adds types to API group
+func addKnownTypes(scheme *runtime.Scheme) error {
+	scheme.AddKnownTypes(SchemeGroupVersion,
```
Comment (runcom) [inline machineconfiguration/v1/register.go]: use `GroupVersion`

File: machineconfiguration/v1/register.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		&ControllerConfigList{},
+		&KubeletConfig{},
+		&KubeletConfigList{},
+		&MachineConfig{},
+		&MachineConfigList{},
+		&MachineConfigPool{},
+		&MachineConfigPoolList{},
+	)
+
+	metav1.AddToGroupVersion(scheme, SchemeGroupVersion)
```
Comment (runcom) [inline machineconfiguration/v1/register.go]: use `GroupVersion`

**Comment by LorbusChris** [comment]
Generating the docs is failing with:
```
Generating swagger type docs for machineconfiguration/v1 at machineconfiguration/v1
panic: interface conversion: ast.Expr is *ast.SelectorExpr, not *ast.Ident

goroutine 1 [running]:
github.com/openshift/api/vendor/k8s.io/apimachinery/pkg/runtime.fieldName(0xc00002db00, 0x0, 0x0)
    /home/chris/Code/go/src/github.com/openshift/api/vendor/k8s.io/apimachinery/pkg/runtime/swagger_doc_generator.go:111 +0x1e5
github.com/openshift/api/vendor/k8s.io/apimachinery/pkg/runtime.ParseDocumentationFrom(0xc000018660, 0x20, 0xc000064930, 0x1, 0x1)
    /home/chris/Code/go/src/github.com/openshift/api/vendor/k8s.io/apimachinery/pkg/runtime/swagger_doc_generator.go:185 +0x22f
main.main()
    /home/chris/Code/go/src/github.com/openshift/api/tools/genswaggertypedocs/swagger_type_docs.go:71 +0x1d3
exit status 2
make: *** [Makefile:25: generate] Error 1
```

Could be some explicit conversion is missing, similar to:
https://github.com/runcom/api/blob/296f4c423703bb8b5b7e122cd4b3fc4153ef1130/build/v1/doc.go#L2

File: machineconfiguration/v1/register.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+var (
+	GroupName     = "machineconfiguration.openshift.io"
+	GroupVersion  = schema.GroupVersion{Group: GroupName, Version: "v1"}
+	schemeBuilder = runtime.NewSchemeBuilder(addKnownTypes, corev1.AddToScheme)
+	// Install is a function which adds this version to a scheme
+	Install = schemeBuilder.AddToScheme
+
+	// SchemeGroupVersion generated code relies on this name
+	// Deprecated
+	SchemeGroupVersion = GroupVersion
```
Comment (LorbusChris) [inline machineconfiguration/v1/register.go:19]: keep the deprecated things around or remove?

Comment (runcom) [reply machineconfiguration/v1/register.go:19]: keep it around, we'll discuss that with openshift/api maintainers surely when you open the official PR tehre

**Comment by LorbusChris** [comment]
next issue:
```
[chris@localhost api]$ make build
go build github.com/openshift/api/...
# github.com/openshift/api/vendor/k8s.io/kubelet/config/v1beta1
vendor/k8s.io/kubelet/config/v1beta1/generated.pb.go:264:13: arguments to copy have different element types: []byte and []string
vendor/k8s.io/kubelet/config/v1beta1/generated.pb.go:2205:26: cannot use make(map[string]string) (type map[string]string) as type map[string][]string in assignment
vendor/k8s.io/kubelet/config/v1beta1/generated.pb.go:2294:33: cannot use mapvalue (type string) as type []string in assignment
make: *** [Makefile:8: build] Error 2
```

Maybe it's this issue: https://github.com/golang/go/issues/23536 ?

**Comment by kikisdeliveryservice** [comment]
the first line 264 error makes sense bc the dAtA is a []byte and v is []string. v is an element of StaticPodURLHeader which is a map[string][]string (see  vendor/k8s.io/kubelet/config/v1beta1/types.go)

**Comment by kikisdeliveryservice** [comment]
for the second error it's hitting:
`m.StaticPodURLHeader = make(map[string]string)` which goes against the StaticPodURLHeader type which is a map[string][]string (noted above)

for 3rd it's hitting:
`m.StaticPodURLHeader[mapkey] = mapvalue` where mapvalue is a string.

Not sure why StaticPodURLHeader is doing this in generated files, but the errors just from poking around seem correct given the def in types.go

Not so familar with the API generation etc tho, but @runcom any idea why this is happening?

**Comment by LorbusChris** [comment]
moving this over to: https://github.com/openshift/api/pull/301

**Pull request closed without merging**